src/main/java/org/motechproject/mots/service/UserService.java

```diff
@@ -26,36 +34,110 @@
-   * Finds users matching all of the provided parameters.
-   * If there are no parameters, return all users.
+   * Finds users matching all of the provided parameters. If there are no parameters, return all
+   * users.
    */
-  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_OR_MANAGE_INCHARGE_USERS_ROLE)
   public Page<User> searchUsers(String username, String email, String name, String role,
       Pageable pageable) throws IllegalArgumentException {
-    return userRepository.search(username, email, name, role, pageable);
+    if (authenticationHelper.getCurrentUser().hasPermission(UserPermission.MANAGE_USERS)) {
+      return userRepository.search(username, email, name, role, pageable);
+    }
+
+    final Optional<UserRole> incharge = getInchargeUserRole();
+    if (incharge.isPresent()) {
+      return userRepository.search(username, email, name, incharge.get().getName(), pageable);
+    }
+
+    return new PageImpl<>(Collections.EMPTY_LIST);
   }
 
-  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
+  /**
+   * Returns UserRoles.
+   */
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_OR_MANAGE_INCHARGE_USERS_ROLE)
   public Iterable<UserRole> getRoles() {
-    return roleRepository.findAll();
+    if (authenticationHelper.getCurrentUser().hasPermission(UserPermission.MANAGE_USERS)) {
+      return roleRepository.findAll();
+    }
+
+    final Optional<UserRole> incharge = getInchargeUserRole();
+    if (incharge.isPresent()) {
+      return Collections.singleton(incharge.get());
+    }
+
+    return Collections.EMPTY_LIST;
   }
 
-  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_ROLE)
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_OR_MANAGE_INCHARGE_USERS_ROLE)
   public User getUser(UUID id) {
     return userRepository.findById(id).orElseThrow(() ->
         new EntityNotFoundException("User with id: {0} not found", id.toString()));
   }
 
+  private Optional<UserRole> getInchargeUserRole() {
+    return roleRepository.findByName(INCHARGE_USER_ROLE);
+  }
+
+  /**
+   * Save User with new encoded password (if it's not blank).
+   *
+   * @param user User to be created.
+   * @return saved User
+   */
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_OR_MANAGE_INCHARGE_USERS_ROLE)
+  public User saveUser(User user, boolean encodeNewPassword) {
+    if (encodeNewPassword) {
+      String newPasswordEncoded = new BCryptPasswordEncoder().encode(user.getPassword());
+      user.setPassword(newPasswordEncoded);
+    }
+
+    return validAndSave(user);
+  }
+
+  /**
+   * Create User with new encoded password.
+   *
+   * @param user User to be created.
+   * @return created User
+   */
+  @PreAuthorize(RoleNames.HAS_MANAGE_USERS_OR_MANAGE_INCHARGE_USERS_ROLE)
+  public User registerNewUser(User user) {
+
+    String newPasswordEncoded = new BCryptPasswordEncoder().encode(user.getPassword());
+    user.setPassword(newPasswordEncoded);
+
+    return validAndSave(user);
+  }
+
+  private User validAndSave(User user) {
```

Comment:
private methods should be after all public methods, move this
also I think you should rename this to validateAndSave